### PR TITLE
Removed end-of-video segments that show a sharp drop in users

### PR DESCRIPTION
**Pull request by HassanJaveed84**
@brianhw  @mulby 
I am not sure if this is the right way to go about it. It basically just looks for a spike in num_users from the end to the start and drops those segments.

### CODE REVIEW

File: edx/analytics/tasks/video.py [hunk 1]
```diff
@@ -405,6 +407,18 @@ def reducer(self, key, viewings):
                 users.add(username)
                 stats['views'] = stats.get('views', 0) + 1
 
+        last_segment = max(usage_map.keys())
+        last_segment_num_users = len(usage_map[last_segment]['users'])
+        for segment in sorted(usage_map.keys()[0:-1], reverse=True):
```
Comment (mulby) [inline edx/analytics/tasks/video.py]: I think you want to apply the slice after the sort, something like:

``` python
sorted(usage_map.keys(), reverse=True)[1:]
```

File: edx/analytics/tasks/video.py [hunk 2]
```diff
@@ -405,6 +407,18 @@ def reducer(self, key, viewings):
...
+            stats = usage_map[segment]
+            current_segment_num_users = len(stats.get('users', []))
+            if last_segment_num_users <= current_segment_num_users * VIDEO_DROPOFF_THRESHOLD:
+                for segment_dropoff in xrange(last_segment, max(usage_map.keys()) + 1):
+                    del usage_map[segment_dropoff]
```
Comment (mulby) [inline edx/analytics/tasks/video.py]: I think this will fail if there is a gap in the data... for example: 100, 102, 103 etc.

We probably want something like this instead: `usage_map.pop(segement_dropoff, None)` or a try-except that catches the key error.

Meta point - I would include this in your test cases when you write unit tests for this code.

File: edx/analytics/tasks/video.py [hunk 3]
```diff
@@ -405,6 +407,18 @@ def reducer(self, key, viewings):
                 users.add(username)
                 stats['views'] = stats.get('views', 0) + 1
 
+        last_segment = max(usage_map.keys())
+        last_segment_num_users = len(usage_map[last_segment]['users'])
+        for segment in sorted(usage_map.keys()[0:-1], reverse=True):
+            stats = usage_map[segment]
+            current_segment_num_users = len(stats.get('users', []))
+            if last_segment_num_users <= current_segment_num_users * VIDEO_DROPOFF_THRESHOLD:
+                for segment_dropoff in xrange(last_segment, max(usage_map.keys()) + 1):
```
Comment (mulby) [inline edx/analytics/tasks/video.py]: You may want to consider caching `max(usage_map.keys())` since it is computed earlier and involves a full traversal of a rather large dictionary.

File: edx/analytics/tasks/video.py [hunk 4]
```diff
@@ -405,6 +407,18 @@ def reducer(self, key, viewings):
                 users.add(username)
                 stats['views'] = stats.get('views', 0) + 1
 
+        last_segment = max(usage_map.keys())
```
Comment (mulby) [inline edx/analytics/tasks/video.py]: I would suggest moving this code to a function and adding a doc string that explains why it is necessary.

**Comment by mulby** [comment]
I like the approach, we should try it out on some real data to see how effective it is at mitigating the issue.

Once we do that we should write unit tests to make sure we cover some of the edge cases here.

File: edx/analytics/tasks/video.py [hunk 5]
```diff
@@ -405,6 +407,18 @@ def reducer(self, key, viewings):
...
+            if last_segment_num_users <= current_segment_num_users * VIDEO_DROPOFF_THRESHOLD:
+                for segment_dropoff in xrange(last_segment, max(usage_map.keys()) + 1):
+                    del usage_map[segment_dropoff]
+                break
+            last_segment_num_users = current_segment_num_users
+            last_segment = segment
+
         # If we don't know the duration of the video, just use the final segment that was
         # actually viewed to determine users_at_end.
         if video_duration == VIDEO_UNKNOWN_DURATION:
```
Comment (brianhw) [inline edx/analytics/tasks/video.py:411]: A good start!  There are two kinds of modifications we want to make to this code.  One is to change how we estimate what the end of the video is.  The second will address how "users_at_end" is calculated below.  I assume this PR is addressing the first.  So I think we only want to do this in cases where we don't know the video's duration.  So I believe the above logic would go below this line.

File: edx/analytics/tasks/video.py (hunk 1 above)
Comment (HassanJaveed84) [inline edx/analytics/tasks/video.py]: Yes, I did, thanks!

Comment (brianhw) [inline edx/analytics/tasks/video.py]: I don't understand the use of [0:-1] here.  You're doing that on the list of keys, before sorting.  Do you know what's at the end of that list?  But I assume the goal is to exclude the last_segment.

File: edx/analytics/tasks/video.py (hunk 5 above)
Comment (HassanJaveed84) [inline edx/analytics/tasks/video.py:411]: Thanks Brian! I'll move the code to a function as Gabe suggested and make the other improvements.

File: edx/analytics/tasks/video.py [hunk 6]
```diff
@@ -38,6 +38,8 @@
 VIDEO_VIEWING_SECONDS_PER_SEGMENT = 5
 VIDEO_VIEWING_MINIMUM_LENGTH = 0.25  # seconds
 
+VIDEO_DROPOFF_THRESHOLD = 0.05 # percentage of users
```
Comment (brianhw) [inline edx/analytics/tasks/video.py]: Going forward, I would suggest that this be made a Luigi parameter, with a default value read from config.

File: edx/analytics/tasks/video.py (hunk 5 above)
Comment (brianhw) [inline edx/analytics/tasks/video.py:411]: It may be a little simplified if we didn't bother with deleting entries.  Instead, we could just calculate a different value for final_segment, and then use final_segment to break from the for-loop below.

File: edx/analytics/tasks/video.py (hunk 6 above)
Comment (brianhw) [inline edx/analytics/tasks/video.py]: Just starting to think about how we might validate this approach (i.e. to look backwards for neighboring segments where the users in the second segment fall to 5% of the first segment).  Again, if this were to be applied only to videos with unknown duration, we could validate by checking what values seem to "work" on videos with known duration.  By applying it in such cases, we could determine how much video we might lose.

**Comment by HassanJaveed84** [comment]
@mulby @brianhw 
Only applies the threshold for videos with unknown duration, also changed the way final_segment is calculated.

**Comment by HassanJaveed84** [comment]
@mulby @brianhw 
Added unit test case for get_final_segment function. I have used PyYAML here to load the input data for the test. 
Made dropoff_threshold a luigi parameter for VideoUsageTask. I think it needs to be passed down from InsertToMysqlAllVideoTask ?

**Comment by mulby** [comment]
@HassanJaveed84 - I'm actually OK with not passing down the drop-off parameter. Allowing it to be set in the config file is good enough for me.

File: config/test.cfg [hunk 7]
```diff
@@ -69,3 +69,6 @@ access_token = acceptance
 [vertica-export]
 credentials = s3://fake/vertica_creds.json
 schema = testing
+
+[videos]
+dropoff_threshold = 0.05
```
Comment (mulby) [inline config/test.cfg:105]: Not sure where you mentioned this, but I agree, I would add this to devstack.cfg as well.

File: edx/analytics/tasks/tests/test_video.py [hunk 8]
```diff
@@ -1013,3 +1016,20 @@ def test_end_view_with_duration(self):
...
             }
         ])
+
+class GetFinalSegmentTest(unittest.TestCase):
+
+    def setUp(self):
+        self.task = VideoUsageTask(dropoff_threshold = 0.10, interval=sentinel.ignored, output_root=sentinel.ignored)
+
+    def test_with_dropoff(self):
+        input = os.path.join(os.path.dirname(__file__), 'fixtures', 'input', 'video_usage_map_with_dropoff.yml')
```
Comment (mulby) [inline edx/analytics/tasks/tests/test_video.py]: I would write a small function that returns the full path given just the filename so that you don't have to repeat this logic in the tests. Or perhaps build most of the path in the setup and just append the filename in the actual test.

Comment (mulby) [inline edx/analytics/tasks/tests/test_video.py]: I think "input" is a builtin function in python, I would use a different name for this variable.

File: edx/analytics/tasks/tests/fixtures/input/video_usage_map_with_dropoff.yml [hunk 9]
```diff
@@ -0,0 +1,45 @@
+0:
```
Comment (mulby) [inline edx/analytics/tasks/tests/fixtures/input/video_usage_map_with_dropoff.yml]: Looking at this input, I'm wondering if you couldn't manufacture "interesting" input that has similar characteristics in the test case. It may actually be more clear what case the input is illustrating.

File: edx/analytics/tasks/tests/test_video.py [hunk 10]
```diff
@@ -1013,3 +1016,20 @@ def test_end_view_with_duration(self):
...
+        with open(input) as data_file:
+            usage_map = yaml.safe_load(data_file)
+        self.assertEqual(self.task.get_final_segment(usage_map), 10)
+
+    def test_without_dropoff(self):
```
Comment (mulby) [inline edx/analytics/tasks/tests/test_video.py:1047]: I'd like to see a test for the case where there is gradual drop-off. Also one where there is a sharp discontinuity but doesn't trigger the cutoff to illustrate the limitations of the algorithm.

This at least indicates to other developers that they are known limitations. I would put in comments describing those cases where the algorithm will do "the wrong thing" with a note about our duration comparison experiment and the results.

File: edx/analytics/tasks/tests/fixtures/input/video_usage_map_with_dropoff.yml (hunk 9 above)
Comment (mulby) [inline edx/analytics/tasks/tests/fixtures/input/video_usage_map_with_dropoff.yml]: To be more clear, I'm wondering if it might be cleaner to programmatically define the input instead of explicitly declaring it in these files.  Something like:

``` python
def generate_segment(num_users, num_views):
    return {
        'users': ['user{}'.format(i) for i in range(num_users)],
        'views': num_views
    }

usage_map = {}
for j in range(6):
    usage_map[j] = generate_segment(20, 15)

# include a couple segments at the end with dramatically fewer users
usage_map[6] = generate_segment(1, 1)
usage_map[7] = generate_segment(1, 1)
```

I like aspects of the current yaml file - it's very clear and explicit, however, I'm nervous that it will be tedious to create interesting edge cases.

@brianhw do you have an opinion on this subject?

File: config/test.cfg (hunk 7 above)
Comment (brianhw) [inline config/test.cfg:105]: and also to our dev/prod config in analytics-secure, when we get to it.

File: edx/analytics/tasks/tests/fixtures/input/video_usage_map_with_dropoff.yml (hunk 9 above)
Comment (brianhw) [inline edx/analytics/tasks/tests/fixtures/input/video_usage_map_with_dropoff.yml]: Sorry to see so much work with the yaml files, but I am more strongly opinionated (or less polite :) about this than Gabe.  I much prefer the use of factory methods over the use of input files for this kind of testing.  It's hard enough to deal with the fixtures in acceptance tests (and I would prefer it if we could do more programmatic work there as well).  Fixtures make sense when one needs to model more complex, real-world type data, and that's not at all the case here.    If you want to get the data values to skip around, then using modulo operators will often give you that.   But I would argue that having a simple base generator to produce a usage_map of a given number of segments, and then others that modify the base, should provide an easier and clearer way to produce the test data.    It looks like there are two sets so far, and the second looks like it just appends to the first.

File: edx/analytics/tasks/tests/test_video.py (hunk 8 above)
Comment (brianhw) [inline edx/analytics/tasks/tests/test_video.py]: Based on my earlier comment, I'm assuming this will be replaced.

File: edx/analytics/tasks/tests/test_video.py (hunk 10 above)
Comment (brianhw) [inline edx/analytics/tasks/tests/test_video.py:1047]: Note that one can get a number of the additional cases by running with the same data but a different threshold.  It's much easier to change the one value to get more coverage.  As Gabe noted before, you can run the same tests over with different thresholds and expected final segment values using ddt.  It's pretty handy, and there are examples elsewhere in the pipeline unit tests.

**Comment by HassanJaveed84** [comment]
@mulby @brianhw 
Thank you for the feedback and the guidance, much appreciated!
I updated the test case, though it looks obscure to me, what do you guys think?

File: edx/analytics/tasks/tests/test_video.py [hunk 11]
```diff
@@ -1013,3 +1016,45 @@ def test_end_view_with_duration(self):
...
+            usage_map[i] = self.generate_segment(20, 20)
+        return usage_map
+
+    @data(
+        (15, 25, 14),
+        (20, 30, 19),
+    )
+    @unpack
+    def test_with_sudden_dropoff(self, start, end, result):
+        for i in range(start, end):
```
Comment (brianhw) [inline edx/analytics/tasks/tests/test_video.py:1043]: Yes, I definitely like this better.  I would just suggest adding a comment on each of these with a brief description of the data.  Or it could be a docstring, as you did for the last one.

File: edx/analytics/tasks/tests/test_video.py [hunk 12]
```diff
@@ -1,9 +1,11 @@
 """Test student engagement metrics"""
 
 import json
+import yaml
```
Comment (brianhw) [inline edx/analytics/tasks/tests/test_video.py]: No longer needed....

File: edx/analytics/tasks/tests/test_video.py [hunk 13]
```diff
@@ -12,6 +14,7 @@
 from edx.analytics.tasks.tests.opaque_key_mixins import InitializeOpaqueKeysMixin, InitializeLegacyKeysMixin
 from edx.analytics.tasks.tests.map_reduce_mixins import MapperTestMixin, ReducerTestMixin
 
+import os
```
Comment (brianhw) [inline edx/analytics/tasks/tests/test_video.py]: No longer needed

File: config/test.cfg (hunk 7 above)
Comment (brianhw) [inline config/test.cfg:105]: I did this already.

**Comment by HassanJaveed84** [comment]
@mulby @brianhw is there anything else that I can improve before we merge it?

**Comment by brianhw** [comment]
Hassan, this looks great!   Next step is to rebase on master, since there seems to be a conflict that will need to be resolved.  And to squash all the intermediate commits in the process so that there's a single commit for this PR.  Let me know if you need help with that!

**Comment by brianhw** [comment]
Oh yeah, for the record:  :+1:

**Comment by HassanJaveed84** [comment]
@brianhw @mulby should I merge it?

**Comment by mulby** [comment]
@HassanJaveed84 - sorry I've been out for the day.

This looks good to me :+1:

You can merge it whenever you are ready.

**Comment by brianhw** [comment]
Yes, feel free to merge.  Or if you would prefer, let us know and we'll hit that button.
